csrc/aes_kwp.cpp

```diff
@@ -0,0 +1,96 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cstdio>
+#include <cassert>
+#include <algorithm> // for std::min
+#include <openssl/aes.h>
+#include <openssl/evp.h>
+#include <openssl/err.h>
+#include "generated-headers.h"
+#include "util.h"
+#include "env.h"
+#include "buffer.h"
+#include "keyutils.h"
+
+using namespace AmazonCorrettoCryptoProvider;
+
+JNIEXPORT int JNICALL Java_com_amazon_corretto_crypto_provider_AesKeyWrapSpi_wrapKey(
+  JNIEnv *pEnv,
+  jclass,
+  jbyteArray keyArray,
+  jbyteArray inputArray,
+  jbyteArray outputArray
+)
+{
+    try {
+        raii_env env(pEnv);
+
+        java_buffer key = java_buffer::from_array(env, keyArray);
+        java_buffer input = java_buffer::from_array(env, inputArray);
+        java_buffer output = java_buffer::from_array(env, outputArray);
+
+        AES_KEY aes_key;
+        SecureBuffer<uint8_t, 32> keybuf;
+        key.get_bytes(env, keybuf.buf, 0, key.len());
+        if (AES_set_encrypt_key(keybuf.buf, key.len()*8, &aes_key) != 0) {
```

Comment:
FYI, this is an expensive call and you might want to save the result if the key is commonly reused. You can see that we do something like that for AES-GCM.

Reply:
that's good to know. i wouldn't have expected this to be an expensive call (it's just DER-decoding a 16-32 byte key?). to keep the initial implementation simple, i think it would be advisable to punt on this now, and revisit it if it becomes a hot spot.

if we were to retain the AES_KEY across wrap/unwrap invocations, we'd track its pointer as a NativeResource, right?

Reply:
AES keys are not DER encoded. However, the AES keyschedule is very expensive. The [Stick Figure Guide to the Advanced Encryption Standard (AES)](http://www.moserware.com/2009/09/stick-figure-guide-to-advanced.html) is a great read. (They call it "key expansion".)

Yes, it would be tracked as a NativeResource.